code/game/objects/structures/lavaland/ore_vent.dm

```diff
@@ -0,0 +1,502 @@
+	)
+	var/wave_timer = 60 SECONDS
+	if(boulder_size == BOULDER_SIZE_MEDIUM)
+		wave_timer = 90 SECONDS
+	else if(boulder_size == BOULDER_SIZE_LARGE)
+		wave_timer = 150 SECONDS
+	COOLDOWN_START(src, wave_cooldown, wave_timer)
+	addtimer(CALLBACK(src, PROC_REF(handle_wave_conclusion)), wave_timer)
+	spawning_mobs = TRUE
+	icon_state = icon_state_tapped
+	update_appearance(UPDATE_ICON_STATE)
+
+/**
+ * Called when the wave defense event ends, after a variable amount of time in start_wave_defense.
+ *
+ * If the node drone is still alive, the ore vent is tapped and the ore vent will begin generating boulders.
+ * If the node drone is dead, the ore vent is not tapped and the wave defense can be reattempted.
+ *
+ * Also gives xp and mining points to all nearby miners in equal measure.
+ */
+/obj/structure/ore_vent/proc/handle_wave_conclusion()
+	SEND_SIGNAL(src, COMSIG_VENT_WAVE_CONCLUDED)
+	COOLDOWN_RESET(src, wave_cooldown)
+	particles = null
+	if(!QDELETED(node)) ///The Node Drone has survived the wave defense, and the ore vent is tapped.
+		tapped = TRUE
+		SSore_generation.processed_vents += src
+		balloon_alert_to_viewers("vent tapped!")
+	else
+		visible_message(span_danger("\the [src] creaks and groans as the mining attempt fails, and the vent closes back up."))
+		icon_state = initial(icon_state)
+		update_appearance(UPDATE_ICON_STATE)
+		return FALSE //Bad end, try again.
+
+	for(var/mob/living/carbon/human/potential_miner in range(7, src)) //Give the miners who are near the vent points and xp.
+		var/mob/living/carbon/human/true_miner = potential_miner
+		var/obj/item/card/id/user_id_card = true_miner.get_idcard(TRUE)
+		if(true_miner?.mind)
+			true_miner.mind.adjust_experience(/datum/skill/mining, MINING_SKILL_BOULDER_SIZE_XP * boulder_size)
+		if(!user_id_card)
+			continue
+		var/point_reward_val = (MINER_POINT_MULTIPLIER * boulder_size) - MINER_POINT_MULTIPLIER // We remove the base value of discovering the vent
+		user_id_card.registered_account.mining_points += point_reward_val
+		user_id_card.registered_account.bank_card_talk("You have been awarded [point_reward_val] mining points for your efforts.")
+	node.pre_escape() //Visually show the drone is done and flies away.
+	add_overlay(mutable_appearance('icons/obj/mining_zones/terrain.dmi', "well", ABOVE_MOB_LAYER, src, GAME_PLANE))
+
+/**
+ * Called when the ore vent is tapped by a scanning device.
+ * Gives a readout of the ores available in the vent that gets added to the description, then asks the user if they want to start wave defense.
+ */
+/obj/structure/ore_vent/proc/scan_and_confirm(mob/user, scan_only = FALSE)
+	if(tapped)
+		balloon_alert_to_viewers("vent tapped!")
+		return
+	if(!COOLDOWN_FINISHED(src, wave_cooldown))
+		balloon_alert_to_viewers("protect the node drone!")
+		return
+	if(!discovered)
+		balloon_alert(user, "scanning...")
+		playsound(src, 'sound/items/timer.ogg', 30, TRUE)
+		if(scan_only)
+			discovered = TRUE
+		if(!discovered)
```

Comment:
Don't like how `discovered` is nested twice. Looks like an opportunity for early return